packages/api/src/controllers/clip.ts

```diff
@@ -0,0 +1,154 @@
+import { validatePost } from "../middleware";
+import { Request, Router } from "express";
+import _ from "lodash";
+import { db } from "../store";
+import { NotFoundError } from "../store/errors";
+import { pathJoin } from "../controllers/helpers";
+import {
+  createAsset,
+  validateAssetPayload,
+  defaultObjectStoreId,
+  catalystPipelineStrategy,
+} from "./asset";
+import { generateUniquePlaybackId } from "./generate-keys";
+import { v4 as uuid } from "uuid";
+import { DBSession } from "../store/db";
+import { fetchWithTimeout } from "../util";
+import { DBStream } from "../store/stream-table";
+
+const app = Router();
+
+app.post("/", validatePost("clip-payload"), async (req, res) => {
+  const playbackId = req.body.playbackId;
+  const userId = req.user.id;
+
+  const id = uuid();
+  let uPlaybackId = await generateUniquePlaybackId(id);
+
+  const content = await db.stream.getByPlaybackId(playbackId); //||
+  //(await db.asset.getByPlaybackId(playbackId));
+
+  let isStream: boolean;
+  if (content && "streamKey" in content) {
+    isStream = true;
+  }
+
+  if (!content) {
+    throw new NotFoundError("Content not found");
+  }
+
+  const user = await db.user.get(content.userId);
+
+  if (!user || userId !== content.userId) {
+    throw new NotFoundError("Content not found");
+  }
+
+  if ("suspended" in content && content.suspended) {
+    throw new NotFoundError("Content not found");
+  }
+
+  let asset = await validateAssetPayload(
+    id,
+    uPlaybackId,
+    content.userId,
+    Date.now(),
+    defaultObjectStoreId(req),
+    req.config,
+    {
+      name: req.body.name || `clip_${uPlaybackId}`,
+    },
+    // TODO: actual type `clip` - currently something's wrong with the types when setting up the schema
+    { type: "directUpload", sourceId: content.id }
+  );
+
+  let url: string;
+
+  let session: DBSession;
+  if (isStream) {
+    if (!content.record) {
+      res.status(400).json({
+        errors: ["Recording must be enabled on a live stream to create clips"],
+      });
+    }
+    let recordingResponse = await getRecordingUrl(content, req);
+    url = recordingResponse.url;
+    session = recordingResponse.session;
+  } else {
```

Comment:
Related to the stream vs session discussion, it feels like this logic should be in some kind of `fetchClipContent` function, and this handler should never even care/access the `Stream` object directly. The stream object is just an implementation detail to get to the session, but clips should conceptually only relate with the sessions themselves.

i.e. on line 28 we can call some function that returns `DBSession | DBAsset` and that's all this API handler cares about

Reply:
Moved into a separate function after talking with max